### PR TITLE
feat(txpool): Dev/geth txpool

### PR DESCRIPTION
**Pull request by itsdevbear**
### Summary by CodeRabbit

Release Notes:

- New Feature: Introduced new interfaces and structs for proposal handling and transaction processing in the Cosmos ABCI package.
- Refactor: Updated contract bytecode and ABI in various contract bindings, reflecting changes in the underlying smart contracts.
- Performance: Optimized loop condition in `DeployAndCallERC20.sol` script to iterate 62226 times instead of 66, potentially improving contract testing efficiency.
- New Feature: Added configuration fields and flags related to the LegacyTxPool module in the Polaris application.
- Refactor: Made significant updates to the `txpool` package, improving transaction handling.
- Test: Introduced several mock implementations for various types in the `txpool` package, enhancing test coverage.
- Chore: Added a `.mockery.yaml` configuration file for generating mock files.
- Bug Fix: Fixed potential issues in Ethereum core and miner packages by modifying interfaces, changing function signatures, and updating method calls.
- Refactor: Updated Keeper struct and its associated functions in the `cosmos/x/evm/keeper` package, improving EVM account balance management.
- Style: Improved readability in `lib/ds/stack/cloneable_stack.go` with a new line break in the function signature.

### CODE REVIEW

**Comment by itsdevbear** [comment]
TODO: 
- base fee
- fix some of the race conditions

File: eth/core/genesis.go [hunk 1]
```diff
@@ -59,12 +54,23 @@ var DefaultGenesis = &core.Genesis{
...
+	// in actual genesis blocks.
+	Number:        0,
+	GasUsed:       0,
+	ParentHash:    common.Hash{},
+	BaseFee:       nil,
+	ExcessBlobGas: nil,
+	BlobGasUsed:   nil,
+
+	// These fields are used for consensus tests. Please don't use them
+	// in actual genesis blocks.
```
Comment (calbera) [inline eth/core/genesis.go:68]: can remove

File: eth/core/host.go [hunk 2]
```diff
@@ -135,27 +134,11 @@ type (
...
-		// well as queued transactions, grouped by account and nonce.
-		Content() (map[common.Address][]*types.Transaction, map[common.Address][]*types.Transaction)
-		// ContentFrom retrieves the data content of the transaction pool, returning the pending
-		// as well as queued transactions of this address, grouped by nonce.
-		ContentFrom(addr common.Address) ([]*types.Transaction, []*types.Transaction)
+		// GetHandler returns a handler that is used to broadcast transactions.
+		GetHandler() Handler
 	}
+
+	Handler interface{}
```
[the remark this answers is not in the document]
Comment (calbera) [reply eth/core/host.go:141]: +1, is this even used?

File: eth/core/processor.go [hunk 3]
```diff
@@ -187,6 +187,7 @@ func (sp *StateProcessor) Finalize(
 
 // BuildAndRegisterPrecompiles builds the given precompiles and registers them with the precompile
 // plugin.
+// TODO: move precompile registration out of the state processor?
```
Comment (calbera) [inline eth/core/processor.go:190]: think this is still necessary every block, as the default geth ones might change. Should probably delete the previous block's rules' default precompiles before registering the current block's rules' default precompiles to be exactly correct

Comment (calbera) [reply eth/core/processor.go:190]: btw will have to add this in geth fork if removing our processor completely, cuz with our injected PrecompileManager system this is now necessary. before hand in upstream geth, it actually checks activePrecompiles every block to effectively get the same result

File: eth/polar/polaris.go [hunk 4]
```diff
@@ -111,14 +113,37 @@ func NewWithNetworkingStack(
...
+func (pl *Polaris) Init() error {
+	pl.miner = miner.New(pl)
+	// eth.miner.SetExtra(makeExtraData(config.Miner.ExtraData))
+
+	var err error
+	legacyPool := legacypool.New(
+		pl.cfg.LegacyTxPool, pl.Blockchain(),
+	)
+
+	pl.txPool, err = txpool.New(big.NewInt(0), pl.blockchain, []txpool.SubPool{legacyPool})
```
Comment (calbera) [inline eth/polar/polaris.go:131]:
```suggestion
	pl.txPool, err = txpool.New(new(big.Int).SetUint64(pl.cfg.LegacyTxPool.PriceLimit), pl.blockchain, []txpool.SubPool{legacyPool})
```
as done in geth:
image

File: eth/polar/polaris.go [hunk 5]
```diff
@@ -147,8 +172,7 @@ func (pl *Polaris) StartServices() error {
 	pl.filterSystem = utils.RegisterFilterAPI(pl.stack, pl.backend, &defaultEthConfig)
 
 	go func() {
-		// TODO: unhack this.
-		time.Sleep(2 * time.Second) //nolint:gomnd // we will fix this eventually.
+		time.Sleep(3 * time.Second) //nolint:gomnd // TODO:fix, this is required for hive...
```
Comment (calbera) [inline eth/polar/polaris.go:175]: is this only required for hive now? then should be a test config flag that only sleeps if test mode

File: cosmos/abci/prepare/prepare.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+	var (
+		selectedTxs  [][]byte
+		totalTxBytes int64
+		totalTxGas   uint64
+	)
+	txs := h.txPoolTransactions()
+	txp, ok := h.polaris.Host().GetTxPoolPlugin().(txpool.Plugin)
+	if !ok {
+		panic("big bad wolf")
```
[the remark this answers is not in the document]
Comment (calbera) [reply cosmos/abci/prepare/prepare.go:73]: lol coderabbit is onto smth here

File: cosmos/abci/prepare/prepare.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		maxBlockGas = b.MaxGas
+	}
+
+	var (
+		selectedTxs  [][]byte
+		totalTxBytes int64
+		totalTxGas   uint64
+	)
+	txs := h.txPoolTransactions()
+	txp, ok := h.polaris.Host().GetTxPoolPlugin().(txpool.Plugin)
```
Comment (calbera) [inline cosmos/abci/prepare/prepare.go:71]: utils.getAs

File: cosmos/abci/prepare/prepare.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (calbera) [reply cosmos/abci/prepare/prepare.go:73]: i didnt know you talk back coderabbitai

**Review by calbera — APPROVED** [review]
some minor nit, but honestly this looks good and I think should be the defacto txpool impl

File: e2e/hive/clients/polard/hive-init.sh [hunk 8]
```diff
@@ -105,5 +110,56 @@
...
+sed -i "s/shanghai-time = .*/shanghai-time = \"$SHANGHAI_TIME\"/" $APP_TOML
+sed -i "s/terminal-total-difficulty = .*/terminal-total-difficulty = \"$TERMINAL_TOTAL_DIFFICULTY\"/" $APP_TOML
+sed -i "s/terminal-total-difficulty-passed = .*/terminal-total-difficulty-passed = $TERMINAL_TOTAL_DIFFICULTY_PASSED/" $APP_TOML
+
+# sed -i 's/"null"/null/g' $APP_TOML
+# sed -i 's/"null"/null/g' $APP_TOML
+# cat "POSTSED"
+# cat $APP_TOML
+cat $APP_TOML
+echo "BINGBONG"
```
[the remark this answers is not in the document]
Comment (calbera) [reply e2e/hive/clients/polard/hive-init.sh:162]: tru say

File: cosmos/abci/prepare/prepare.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (calbera) [reply cosmos/abci/prepare/prepare.go:73]: ur cracked

File: eth/polar/polaris.go (hunk 5 above)
Comment (itsdevbear) [reply eth/polar/polaris.go:184]: handle in later pr

**Pull request merged**